### PR TITLE
Fix bug in polygon_contains_point function

### PR DESCRIPTION
**Pull request by MFraters**
Fix bug in polygon_contains_point function, when a point is exactly on a point_list point. Also adds some brackets for clarity.

### CODE REVIEW

**Comment by tjhei** [comment]
It is hard to tell from this diff, if this is correct now. Do you have an example where this failed? /run-tests

**Comment by MFraters** [comment]
I agree that it is hard to tell from the diff. The bugfix is actually only line 147. I could come with a specific example, but when you compare this code with the code where this was derived from: http://geomalgorithms.com/a03-_inclusion.html you can see that the line with // an upward crossing now matches.

**Comment by tjhei** [comment]
Your test is currently failing: http://www.math.clemson.edu/~heister/cib/aspect-8.4.1/39a93080754aca604798826a107e4bcfbe22c35c/changes-clang.diff Do you understand why and can you fix this please?

**Comment by MFraters** [comment]
Hey Timo,

Yes, I understand why it is happening, but I am not really happy with it. When I designed this, I wanted it to consider everything exactly on the boundary as inside. The default algorithm considers points on the boundaries sometimes to be inside and sometimes to be outside the polygon. By changing the code I 'fixed' this. But recently I realised that this 'fix' actually caused a bug, namely, when I define a polygon with two points parallel to x or y axis, all points on the same line are considered to be inside the polygon. An extended test result of that looks like this:

```
Loading shared library <./libprm_polygon_in_out.so>
Testing prm polygon plugin with the following parameters: 100 > 0,0;5,0;5,5;0,5 & -100 > 10,10;10,15;15,15
Topo at (-1,-1) = 0
Topo at (0,0) = 100
Topo at (0,5) = 100
Topo at (5,0) = 100
Topo at (5,5) = 100
Topo at (5,5.01) = 0
Topo at (1,1) = 100
Topo at (12.5,12) = 0
Topo at (11.5,12) = -100
Topo at (15,10) = 0
Topo at (14,10) = 0
Topo at (15,11) = 0
Topo at (5,6) = 0
Topo at (6,5) = 100
Topo at (0,-1) = 100
Topo at (-1,0) = 100
Topo at (-1,5) = 100
Topo at (0,6) = 100
Topo at (5,-1) = 100
Topo at (6,0) = 100
Topo at (6,-1) = 0
Topo at (16,16) = 0
```

 You can see that the corner points are indeed included (0,0), (0,5),(5,0),(5,5), but also points which should be included like (6,5) and (0,-1). I will see if I can change the code in such a way that it does both as intended, but otherwise is the default algorithm better then what is now implemented.

I hope this clears up the problem a bit :)

**Comment by MFraters** [comment]
Ok, I think I have fixed it in a way so that it does what it promises, and doesn't have the bug. I greatly expanded the test for this to include all kind of critical points.

File: source/utilities.cc [hunk 1]
```diff
@@ -236,35 +236,80 @@ namespace aspect
...
+                    {
+                      // P left of  edge
+                      ++wn;            // have  a valid up intersect
+                    }
+                  else if ( is_left == 0)
+                    {
+                      // The point is exactly on the infinite line.
+                      // determine if it is on the segment
+                      const double dot_product = (point[0] - point_list[j][0]) * (point_list[i][0] - point_list[j][0])
+                                                 + (point[1] - point_list[j][1]) * (point_list[i][1] - point_list[j][1]);
```
Comment (bangerth) [inline source/utilities.cc]: You can write this more concisely as `(point - point_list[j]).norm_square()`.

File: source/utilities.cc [hunk 2]
```diff
@@ -236,35 +236,80 @@ namespace aspect
...
+
+                      if (dot_product >= 0)
+                        {
+                          const double squaredlength = (point_list[i][0] - point_list[j][0]) * (point_list[i][0] - point_list[j][0])
+                                                       + (point_list[i][1] - point_list[j][1]) * (point_list[i][1] - point_list[j][1]);
```
Comment (bangerth) [inline source/utilities.cc]: and this as `(point_list[i] - point_list[j]).norm_square()`.

File: source/utilities.cc [hunk 3]
```diff
@@ -236,35 +236,80 @@ namespace aspect
...
+                      // P right of  edge
+                      --wn;            // have  a valid down intersect
+                    }
+                  else if ( is_left == 0)
+                    {
+                      // This code is to make sure that the boundaries are included in the polygon.
+                      // The point is exactly on the infinite line.
+                      // determine if it is on the segment
+                      const double dot_product = (point[0] - point_list[j][0]) * (point_list[i][0] - point_list[j][0])
+                                                 + (point[1] - point_list[j][1]) * (point_list[i][1] - point_list[j][1]);
```
Comment (bangerth) [inline source/utilities.cc]: same here

File: source/utilities.cc [hunk 4]
```diff
@@ -236,35 +236,80 @@ namespace aspect
...
+
+                      if (dot_product >= 0)
+                        {
+                          const double squaredlength = (point_list[i][0] - point_list[j][0]) * (point_list[i][0] - point_list[j][0])
+                                                       + (point_list[i][1] - point_list[j][1]) * (point_list[i][1] - point_list[j][1]);
```
Comment (bangerth) [inline source/utilities.cc]: and here

**Comment by bangerth** [comment]
I have to admit that I don't understand the algorithm (nor have tried too hard). But I will note that the attempt to definitively rule a point on a segment of a polygon inside/outside the polygon is going to be futile: For this, you would have to do comparisons in exact arithmetic, but all you have a floating point arithmetic. As a consequence, the best you can hope for is to determine whether a point is "safely inside" or "safely outside" a polygon, where "safely" means "within round-off accuracy, i.e., around 1e-15 times the diameter of the polygon". Everything that is within this tolerance of the boundary can not be definitely decided.

File: source/utilities.cc (hunk 1 above)
Comment (MFraters) [inline source/utilities.cc]: This doesn't work, because it doesn't account for point_list[i]

File: source/utilities.cc (hunk 2 above)
Comment (MFraters) [inline source/utilities.cc]: This works!

**Comment by MFraters** [comment]
I agree that the addition of defining what happens on the boundary points is more a theoretical thing for the application I want to use it for, than that it it likely to really make a significant difference in practice. Although I must say that the bug showed me that a point being exactly on one of the axes actually happened often enough to mess up my model with dotted sheets of hot material throughout my model. 

So to fix that bug is the most important for me. It has my preference to 'make sure' that all the boundary points are included, but if it has your preference, I am fine simplifying the code and changing the description that the boundaries of this function can be either in or out, depending on what 'corner' of the polygon you are in.

File: source/utilities.cc (hunk 1 above)
Comment (bangerth) [inline source/utilities.cc]: Oh, I misread then. But you can write it as `(point - point_list[j])*(point_list[i] - point_list[j])` then.

**Comment by bangerth** [comment]
I have no preference one way or the other. I would document that the function gives no guarantees either way. In practice, if your model depends on one behavior or the other, then the model is probably not well posed. For example, it may be better to describe the model with a tapering function that has some kind of smooth transition.

You're the master of that code, so I will defer to you how you want to implement it (subject to the small comments I had).

**Comment by MFraters** [comment]
ok, I have edited the description of the point in polygon to state that that there are no guarantees for the borders.

File: source/utilities.cc [hunk 5]
```diff
@@ -230,41 +230,85 @@ namespace aspect
        *
        * The main functional difference between the original code and this
        * code is that all the boundaries are condidered to be inside the
-       * polygon.
+       * polygon. One should ofcourse realise that with floating point
```
Comment (bangerth) [inline source/utilities.cc]: ofcourse -> of course
realise -> realize

**Comment by MFraters** [comment]
I have addressed the comments and squashed it into two commits, one for the code and one test.

**Pull request merged**